src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -461,6 +461,23 @@ not_ (ECompositeKey _) = throw (CompositeKeyErr NotError)
 (*.)  :: PersistField a => SqlExpr (Value a) -> SqlExpr (Value a) -> SqlExpr (Value a)
 (*.)  = unsafeSqlBinOp " * "
 
+-- | @BETWEEN@.
+--
+-- /Since: 3.0.0/
+between :: PersistField a => SqlExpr (Value a) -> (SqlExpr (Value a), SqlExpr (Value a)) -> SqlExpr (Value Bool)
+a@ERaw{} `between` (ERaw p1 f, ERaw p2 g) =
+  unsafeSqlBinOp " BETWEEN " a $ ERaw Never $ \x ->
+    let (v1, fv) = f x
+        (v2, gv) = g x
+    in  (parensM p1 v1 <> " AND " <> parensM p2 v2, fv ++ gv)
+a `between` (b, c)                        =
+  let construct :: PersistField a => SqlExpr (Value a) -> SqlExpr (Value a)
+      construct (ERaw p f) = throw $ CompositeKeyErr BetweenError
```

Comment:
I don't think this is necessary either -

`ERaw` happens when we do `unsafeSqlFunction`. However, we have a guarantee from the types that all the expressions are `SqlExpr (Value a)`. This means that, as far as Haskell is concerned, we have the same type for everything.

If it is a `ECompositeKey` on one side, then it's going to either be `ECompositeKey` on the other sides, *or* it is going to be a `SqlExpr` that represents the same type of thing.

Like we might call this with: `SqlExpr (Value Rational)` or `SqlExpr (Value UserCompositeKey)`. But if we do, then we know that we'll get it right.

So, I think for this branch, we can just write:

```
a `between` (b, c) = a >=. b &&. a <=. c
```

Does this pass all tests?

Reply:
If we use that directly it fails because `>=.`, `<=.` and `&&.`  depend on `unsafeSqlBinOp` and it [doesn't support composite keys](https://github.com/bitemyapp/esqueleto/blob/a0274e377b7f0eaa4eaf1d2f254e2d803be9ed7e/src/Database/Esqueleto/Internal/Internal.hs#L1650) :disappointed_relieved:

Reply:
Also, while I was doing tests and while writing queries I normally use `val` function but that just passes every value for the placeholder `?` even composite keys making them `ERaw` values, making errors at runtime.

For example this will also fail:
```haskell
select $
  from $ \p -> do
    where_ $
      p ^. PointId
        `between`
           ( val $ PointKey 1 2
           , val $ PointKey 5 6)
```

Reply:
Right, `val` does `ERaw` - you'd get an `ECompositKey` from projecting from a point, like `p ^. PointId`

I think fixing the `unsafeSqlBinOp` to support composite keys makes sense to do here.

Reply:
That sounds good to me :smile: but when we have the scenario that I described before, meaning that for `unsafeSqlBinOp` to handle `ERaw` and `ECompsiteKey` or backward, should it keep failing or how it should be handled?

Reply:
There's pretty much no reason to fail that - we should just render the two arguments on either side. The function is already labeled as `unsafe` so failing at runtime with a SQL syntax error isn't outside the realm of possibility.

Reply:
Great! I'll work on that then, thanks for the help :smile: